proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+
+The Pinniped Supervisor currently supports OIDC-compliant identity providers and
+LDAP identity providers (including Active Directory). Unfortunately, GitHub does not offer
+either of these protocols for authenticating users. GitHub uses a slightly customized version
+of OAuth 2.0 for authenticating users, so Pinniped will need to add new features to support
+using GitHub as an identity provider.
+
+## Terminology / Concepts
+
+### GitHub OAuth 2.0 Clients
+
+For web browser-based authentication, GitHub supports two different types of OAuth 2.0 clients (compared in
+[this doc](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/deciding-when-to-build-a-github-app)
+and [this doc](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/differences-between-github-apps-and-oauth-apps)):
+
+- GitHub Apps (newer)
+- OAuth Apps (we will call these "GitHub OAuth Apps" below to differentiate them from the generic concept of an OAuth
+  2.0 client)
+
+Both flavors of Apps allow delegated authentication and authorization, but GitHub Apps are newer and allow
+GitHub Organization administrators to have more control over the management of the apps.
+
+Astute readers may note that OAuth 2.0 is typically used only for delegated authorization, not authentication.
+GitHub's REST API enables the use of OAuth 2.0 for authentication by offering endpoints that can be used to
+learn about the user's identity from their OAuth 2.0 access token (see below).
+
+### GitHub PATs
+
+For CLI-based (no browser required) authentication, GitHub supports two different types of
+Personal Access Tokens (PATs) (compared in
+[this doc](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens)):
+
+- Fine-grained PATs (newer)
+- Classic PATs
+
+Both flavors of PATs allow authenticating to the GitHub API, but Fine-grained PATs are newer and allow
+GitHub organization administrators to have more control over how the PAT can be used to access API resources
+related to the organization.
+
+Note that at the time of writing this document, Fine-grained PATs are a Beta feature of GitHub, and are not
+enabled by default for use on an organization. The organization admin must opt in to allowing them, and may
+also choose to opt-out of allowing classic PATs.
+
+### GitHub's REST API
+
+GitHub offers a REST API which can be used to find out about, among other things, the
+[identity of the currently authenticated user](https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-the-authenticated-user)
+(`/user`), the
+[organization memberships of the currently authenticated user](https://docs.github.com/en/rest/orgs/orgs?apiVersion=2022-11-28#list-organizations-for-the-authenticated-user)
+(`/user/orgs`), the
+[team memberships of the currently authenticated user](https://docs.github.com/en/rest/teams/teams?apiVersion=2022-11-28#list-teams-for-the-authenticated-user)
```

Comment:
I was thinking if having a large number of groups from a large number of GH orgs would create a burden on Pinniped admins to create the necessary transformations to bridge the gaps between existing GitHub orgs/teams and their existing clusters (probably pre-existing RBAC group names).

One thought that occurred to me is that we could allow a fourth type of `Transformation` specifically for group mappings. I feel like this might be easier to clearly structure for a human than a CEL expression.

One example:
```yaml
- type: groupmappings/v1
  unmapped-behavior: passthrough|drop # Should this transformation pass through or drop unmapped GitHub team names?
  mappings:
    name-of-kube-group-1:
    - github-project-1/github-team-name-1
    - github-project-2/github-team-name-2
    other-kube-group-name:
    - github-project-1/github-team-name-3
    - github-project-3/github-team-name-4
```

Another example, flipping the group mapping. This is not as human-readable but I think it is valid YAML.

```yaml
- type: groupmappings/v1
  unmapped-behavior: passthrough|drop # Should this transformation pass through or drop unmapped GitHub team names?
  mappings:
    [github-project-1/github-team-name-1, github-project-2/github-team-name-2]: name-of-kube-group-1
    [github-project-1/github-team-name-3, github-project-3/github-team-name-4]: other-kube-group-name
```

Reply:
This idea should not be implemented as part of this proposal and can wait for the necessity to arise based on feedback.

Reply:
This is interesting.  I am also inclined to think we should stick to the CEL impl to begin with and see if we do get any feedback or feature requests.  The groups structure for Github is rather odd. 🤔